libsolidity/interface/FileReader.h

```diff
@@ -0,0 +1,105 @@
+/// also used in the solc language server where solc is a long running process.
+class FileReader
+{
+public:
+	using StringMap = std::map<SourceUnitID, SourceCode>;
+	using PathMap = std::map<SourceUnitID, boost::filesystem::path>;
+	using FileSystemPathSet = std::set<boost::filesystem::path>;
+
+	/// Constructs a FileReader with a base path and a set of allowed directories that
+	/// will be used when requesting files from this file reader instance.
+	explicit FileReader(
+		boost::filesystem::path _basePath = {},
+		FileSystemPathSet _allowedDirectories = {}
+	):
+		m_basePath(std::move(_basePath)),
+		m_allowedDirectories(std::move(_allowedDirectories)),
+		m_sourceCodes(),
+		m_sourceCodeOrigin()
+	{}
+
+	void setBasePath(boost::filesystem::path _path) { m_basePath = std::move(_path); }
+	std::string basePath() const noexcept { return m_basePath.generic_string(); }
+
+	void allowDirectory(boost::filesystem::path _path) { m_allowedDirectories.insert(std::move(_path)); }
+	FileSystemPathSet const& allowedDirectories() const noexcept { return m_allowedDirectories; }
+
+	StringMap const& sourceCodes() const noexcept { return m_sourceCodes; }
+
+	/// Retrieves the list of mappings between source unit ID and their representative file system path.
+	PathMap const& origins() const noexcept { return m_sourceCodeOrigin; }
+
+	/// Retrieves the source code for a given source unit ID.
+	SourceCode const& sourceCode(SourceUnitID const& _sourceUnitID) const { return m_sourceCodes.at(_sourceUnitID); }
+
+	/// Resets all sources to the given map of source unit ID to source codes.
+	void setSources(StringMap _sources);
+
+	/// Adds the source code for a given file system path.
+	///
+	/// The source unit ID will be equivalent to the source unit ID.
+	void setSource(boost::filesystem::path _fspath, SourceCode _source);
+
+	/// Adds the source code for a given source unit ID.
+	void setSource(SourceUnitID sourceUnitID, SourceCode _source);
+
+	/// Adds given source, imported as @p _path with filesystem path @p _fspath.
+	void setSource(SourceUnitID _sourceUnitID, std::optional<boost::filesystem::path> _fspath, SourceCode _source);
+
+	/// Reads a given file at @p _path of kind @p _kind from the local filesystem and returns the result.
+	/// @p _kind must always be passed as "source".
+	frontend::ReadCallback::Result readFile(std::string const& _kind, std::string const& _path);
+
+	frontend::ReadCallback::Callback reader()
+	{
+		return [this](std::string const& _kind, std::string const& _path) { return readFile(_kind, _path); };
+	}
+
+private:
+	/// Base path, used for resolving relative paths in imports.
+	boost::filesystem::path m_basePath;
+
+	/// list of allowed directories to read files from
+	FileSystemPathSet m_allowedDirectories;
+
+	/// map of input files to source code strings
+	StringMap m_sourceCodes;
+
+	/// map of input file names to full path names suitable for file://-URIs
+	PathMap m_sourceCodeOrigin;
```

Comment:
I'm still not convinced that this is the right way to tell LSP where the files reside. It does not take base path into account, not all overloads of `setSource()` fill it and `setSources()` can just casually clear it. It does not provide access to files not loaded by `FileReader`.

You suggested that you could move it to the LSP PR. I said it's not necessary but now I think it's actually a good idea because for me it's the only thing holding back this PR. I'm not sure if I'll be reviewing the LSP PR so how about putting it in a PR of its own? We could then discuss it in detail. I think this functionality should really sit in `CompilerStack`, not here.

Reply:
ok. removed.